src/build_context.rs

```diff
@@ -1,133 +1,104 @@
-        self.reusable.contains_key(ref_) || self.slots.iter().any(|slot| slot.slot_ref == ref_)
-    }
-
-    /// check if a ref is used elsewhere in the schema
-    pub fn ref_used(&self, ref_: &str) -> bool {
-        self.used_refs.contains(ref_)
-    }
-
-    /// check if a ref is used within a given schema
-    pub fn ref_used_within(&self, schema_dict: &PyAny, ref_: &str) -> PyResult<bool> {
-        check_ref_used(schema_dict, ref_)
-    }
-
-    /// add a validator/serializer to `reusable` so it can be cloned and used again elsewhere
-    pub fn store_reusable(&mut self, ref_: String, val_ser: T) {
-        self.reusable.insert(ref_, val_ser);
+    fn get_ref_from_id(&self, ref_id: usize) -> &String {
+        // We can unwrap here because nothing should ever have a ref_id unless they've inserted it.
+        self.refs
+            .iter()
+            .find_map(|(key, &val)| if val == ref_id { Some(key) } else { None })
+            .unwrap()
     }
 
-    /// First of two part process to add a new validator/serializer slot, we add the `slot_ref` to the array,
-    /// but not the actual `validator`/`serializer`, we can't add that until it's build.
-    /// But we need the `id` to build it, hence this two-step process.
-    pub fn prepare_slot(&mut self, slot_ref: String) -> PyResult<usize> {
-        let id = self.slots.len();
-        let slot = Slot {
-            slot_ref,
-            op_val_ser: None,
-        };
-        self.slots.push(slot);
-        Ok(id)
+    pub fn get_ref_id(&mut self, reference: String) -> usize {
+        match self.refs.entry(reference) {
+            Entry::Occupied(entry) => *entry.get(),
+            Entry::Vacant(entry) => {
+                let id = self.definitions.len();
+                entry.insert(id);
+                self.definitions.push(None);
+                id
+            }
+        }
     }
 
-    /// Second part of adding a validator/serializer - we update the slot to include a validator
-    pub fn complete_slot(&mut self, slot_id: usize, val_ser: T) -> PyResult<()> {
-        match self.slots.get(slot_id) {
-            Some(slot) => {
-                self.slots[slot_id] = Slot {
-                    slot_ref: slot.slot_ref.clone(),
-                    op_val_ser: Some(val_ser),
-                };
-                Ok(())
-            }
-            None => py_err!("Slots Error: slot {} not found", slot_id),
+    pub fn add_definition(&mut self, ref_id: usize, value: T) -> PyResult<()> {
+        match self.definitions.get_mut(ref_id) {
+            Some(current) => match current {
+                Some(_) => {
+                    let reference = self.get_ref_from_id(ref_id);
+                    py_err!(format!("Duplicate ref: `{reference}`"))
+                }
+                None => {
+                    current.replace(value);
+                    Ok(())
+                }
+            },
+            None => unreachable!(),
```

Comment:
I would prefer an python error here - just that rust panicking can cause seg faults in python that are hard to track down - should be asy here since you return a `PyResult`.

Reply:
made it a python error